vlm/src/main/scala/geotrellis/contrib/vlm/gdal/VSIPath.scala

```diff
@@ -0,0 +1,150 @@
+package geotrellis.contrib.vlm.gdal
+
+
+import java.net.URI
+
+
+object Schemes {
+  final val FTP = "ftp"
+  final val HTTP = "http"
+  final val HTTPS = "https"
+
+  final val TAR = "tar"
+  final val ZIP = "zip"
+  final val GZIP = "gzip"
+
+  final val FILE = "file"
+
+  final val S3 = "s3"
+
+  final val GS = "gs"
+
+  final val WASB = "wasb"
+  final val WASBS = "wasbs"
+
+  final val HDFS = "hdfs"
+
+  final val FILE_TYPE_TO_SCHEME =
+    Map(
+      TAR -> TAR,
+      "tgz" -> TAR,
+      ZIP -> ZIP,
+      "kmz" -> ZIP,
+      "ods" -> ZIP,
+      "xlsx" -> ZIP,
+      GZIP -> GZIP
+    )
+}
+
+
+object VSIPath {
+  import Schemes._
+
+  def apply(path: String): String =
+    apply(new URI(path))
+
+  def apply(uri: URI): String = {
+    val path: String = uri.toString
+    val scheme: String = uri.getScheme
+
+    val isWindows: Boolean =
+      System.getProperty("os.name").toLowerCase == "win"
+
+    def schemeForFileType(targetPath: String): String =
+      FILE_TYPE_TO_SCHEME
+        .get(targetPath.substring(targetPath.lastIndexOf(".") + 1)) match {
+        case Some(scheme) => s"/vsi$scheme/"
+        case None => ""
+      }
+
+    def formatPath(targetURI: URI): String = {
+      lazy val uriPath = targetURI.getPath
+
+      targetURI.getScheme match {
+        case (FTP | HTTP | HTTPS) =>
+          s"/vsicurl/${targetURI.toString}"
+        case S3 =>
+          s"/vsis3/${targetURI.getAuthority}$uriPath"
+        case GS =>
+          s"/vsigs/${targetURI.getAuthority}$uriPath"
+        case (WASB | WASBS) =>
+          val azurePath =
+            if (uriPath.startsWith("/")) uriPath.drop(1) else uriPath
+
+          s"/vsiaz/${targetURI.getUserInfo}/$azurePath"
+        case HDFS =>
+          s"/vsihdfs/${targetURI.toString}"
+        case FILE =>
+          if (isWindows)
+            targetURI.toString.split("file:/").tail.head
```

Comment:
keep in mind that both `tail` and `head` operations are _unsafe_. It means if for some reason it won't be possible to extract tail / head it will throw an exception that can be pretty confusing.

Reply:
That's a good point. I could probably figure out a more safe way of extracting the right info from the `uri`.